overlord/snapstate/handlers.go

```diff
@@ -2254,6 +2254,10 @@ func (m *SnapManager) doDiscardSnap(t *state.Task, _ *tomb.Tomb) error {
 		t.Errorf("cannot remove snap file %q, will retry in 3 mins: %s", snapsup.InstanceName(), err)
 		return &state.Retry{After: 3 * time.Minute}
 	}
+	if err = m.backend.RemoveSnapMountUnits(snapsup.placeInfo(), nil); err != nil {
```

Comment:
this should be done under the `if len(snapst.Sequence) == 0` otherwise it will happen again even before the snap is fully removed.  The test need reflect this. Also not sure if you added a test about checking that refreshes alone don't remove the units.

Reply:
I updated this. I also fixed the spread test, to make sure that mount units are not removed when installing an updated version of a snap.